tokio-postgres/src/config.rs

```diff
@@ -85,7 +89,9 @@ pub enum Host {
 /// * `options` - Command line options used to configure the server.
 /// * `application_name` - Sets the `application_name` parameter on the server.
 /// * `sslmode` - Controls usage of TLS. If set to `disable`, TLS will not be used. If set to `prefer`, TLS will be used
-///     if available, but not used otherwise. If set to `require`, TLS will be forced to be used. Defaults to `prefer`.
+///     if available, but not used otherwise. If set to `require`, `verify-ca`, or `verify-full`, TLS will be forced to
+///     be used. Defaults to `prefer`. Note that for modes `verify-ca` and `verify-full`, it's up to the user to configure
+///     the SSL stream to respect the desired configuration (e.g. verification of certs, hostname verification).
```

Comment:
This note seems like a pretty big footgun.

Reply:
@sfackler Any opinion here? (Sorry to ping you again, please let me know if I should just wait.)

Reply:
I'm not sure how doable it would be to adjust `MakeTlsConnect` and/or `TlsConnect` implementations to apply extra configuration after they were initially constructed.

Reply:
What about introducing convenience constructors that create `MakeTlsConnect` and/or `TlsConnect` with correct configurations (e.g. like [here](https://github.com/sfackler/rust-postgres/pull/774/files#diff-de463e7f8bf25884492679e414d50bf79214f9c353388b1380d8969551661038R104))? The checks (item 3 in the comment above) would guard against wrong configurations and the constructor would be easy to use, e.g. `make_tls(config: &Config)`.

Reply:
But if a user does not realize they need to manually parse their config string up front and use that to create the connector things will silently not work as expected.

Reply:
> things will silently not work as expected

In item 3 [above](https://github.com/sfackler/rust-postgres/pull/774#discussion_r635199367), the idea was that it will actually fail (and not be silently ignored). Would that be an acceptable solution?

If yes, I could follow-up with an implementation. If no, I think we should close the PR and the linked issue.

Reply:
@sfackler looking into this more - it is fairly complicated to pre-parse sslmode because the connection string could come in many different forms (e.g. url-like or space separated), so every implementation would have to duplicate the whole config parsing code without any easy way to extend what you already have.

I propose `pg::config::Config::from_str(conn_str)` to support all ssl modes as defined by psql, but `PostgresConnectionManager` will continue to only support `disable/prefer/require`.  If other ssl_mode is given, `connect_tls` will return an error.

This way, the code using this library will have consistent behavior:
* if the connection string has `sslmode=verify-full`, this library will parse it, but `connect_tls` will give a TLS error as `kind=tls` or `kind=authentication` (?)
* if the user of the lib want to support it, they will have to handle it while constructing the tls connector, and will modify config with `config.clone_as_tls_required()` -- which will keep all of the original config vals but replace ssl mode with `required`.

The benefit of this approach is that all existing code will continue working as before, and if the user doesn't handle new ssl modes, they will continue getting an error - its just that that error will happen a bit later in the process.